samples/matter/common/src/matter_init.h

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#pragma once
+
+#include <DeviceInfoProviderImpl.h>
+#include <app/clusters/network-commissioning/network-commissioning.h>
+#include <app/server/Server.h>
+#include <lib/core/Optional.h>
+#include <lib/support/Variant.h>
+#include <platform/PlatformManager.h>
+
+#ifdef CONFIG_CHIP_WIFI
+#include <platform/nrfconnect/wifi/NrfWiFiDriver.h>
+#endif
+
+#if CONFIG_CHIP_FACTORY_DATA
+#include <platform/nrfconnect/FactoryDataProvider.h>
+#else
+#include <platform/nrfconnect/DeviceInstanceInfoProviderImpl.h>
+#endif
+
+namespace Nordic
+{
+namespace Matter
+{
+	using CustomInit = CHIP_ERROR (*)(void);
+
+	/** @brief Matter initialization data.
+	 *
+	 * This structure contains all user specific implementations of Matter interfaces
+	 * and custom initialization callbacks that must be initialized in the Matter thread.
+	 */
+	struct InitData {
+		/** @brief Pointer to the user provided NetworkCommissioning instance. */
+#ifdef CONFIG_CHIP_WIFI
+		chip::app::Clusters::NetworkCommissioning::Instance *mNetworkingInstance{ &sWiFiCommissioningInstance };
+#else
+		chip::app::Clusters::NetworkCommissioning::Instance *mNetworkingInstance{ nullptr };
+#endif
+		/** @brief Pointer to the user provided custom server initialization parameters. */
+		chip::CommonCaseDeviceServerInitParams *mServerInitParams{ &sServerInitParamsDefault };
```

Comment:
This doesn't have to be `CommonCase` object. Any `ServerInitParams` will do, right?

Reply:
The `ServerInitParams` API does not include `InitializeStaticResourcesBeforeServerInit` which is CommonCase specific... so again, it looks similar to the FactoryDataProxy problem. IMO It's a pity that instead of always using pure Matter interfaces sometimes we need to know implementation specific details. On the other hand CommonCase is explicitly discouraged to be used in the final product firmware, so it's a bit confusing. I wonder if our customers are aware of that :)
https://github.com/nrfconnect/sdk-connectedhomeip/blob/master/src/app/server/Server.h#L153
Transitioning to handling all CommonCase internals in the application now seems a bit overwhelming.

Reply:
Well, I think in this case we can somehow get rid off the `CommonCase` but we can clean that later*. But overall, yes, it looks like default initializers make it difficult to use default objects with custom init. 

\* in some cases we could do:
```
struct InitData
{
  ServerInitParams* mServerInitParams = [] {
    sDefaultServerInitParams.InitializeStaticResourcesBeforeServerInit();
    return &sDefaultServerInitParams
  }();
};
```
But I'm not sure if I'm not proposing monstrosity that would only confuse people :).

Reply:
I have created a task for that, to revisit the usage of CommonCase in NCS Matter samples:
https://nordicsemi.atlassian.net/browse/KRKNWK-18274